src/metabase/lib/metadata.cljc

```diff
@@ -0,0 +1,166 @@
+
+;;; TODO -- consider whether this should dispatch on Field type as well so it's easier to write different
+;;; implementations for Field IDs vs String names vs Keywords etc
+
+(defmulti ^:private field-metadata*
+  "Implementation for [[field-metadata]]."
+  {:arglists '([metadata table field])}
+  (fn [metadata _table _field]
+    (lib.dispatch/dispatch-value metadata)))
+
+(defmethod field-metadata* :metadata/database
+  [database-metadata table-name-or-id field]
+  (assert (some? table-name-or-id)
+          (i18n/tru "Table name or ID is required to fetch a Field from Database metadata"))
+  (field-metadata* (table-metadata database-metadata table-name-or-id) nil field))
+
+(defmethod field-metadata* :metadata/table
+  [table-metadata _table field-id-or-name]
+  (or (some (if (integer? field-id-or-name)
+              (fn [field]
+                (when (= (:id field) field-id-or-name)
+                  field))
+              (fn [field]
+                (when (= (:name field) field-id-or-name)
+                  field)))
+            (:fields table-metadata))
+      ;;; TODO -- not sure it makes sense this that would throw an error while [[field-metadata]] is currently allowed
+      ;;; to return `nil`, and while the other implementations of this do not
+      (throw (ex-info (i18n/tru "Could not find Field {0} in Table {1}"
+                                (pr-str field-id-or-name)
+                                (pr-str (:name table-metadata)))
+                      {:metadata table-metadata
+                       :field    field-id-or-name}))))
+
+;;; TODO -- should this throw an error?
+(defmethod field-metadata* :metadata/results
+  [results-metadata _table field-id-or-name]
+  (some (if (integer? field-id-or-name)
+          (fn [field]
+            (when (= (:id field) field-id-or-name)
+              field))
+          (fn [field]
+            (when (= (:name field) field-id-or-name)
+              field)))
+        (:columns results-metadata)))
+
+(defmethod field-metadata* :mbql/query
+  [query table field]
+  (field-metadata* (:lib/metadata query) table field))
+
+;;; TODO -- a little weird that this can return `nil` but [[table-metadata]] can't... but we have real cases where it
+;;; makes sense that this might return `nil`, unlike for Tables
+
+;;; TODO -- what about nested Fields??
+(mu/defn field-metadata :- [:maybe FieldMetadata]
+  "Get metadata for a specific Field from some `metadata` source, which might be Database metadata, Table metadata, or
+  source query/results metadata. `table` is optional; if you specify it and pass Database metadata,
+  this will first find the appropriate [[table-metadata]] for that Table, then find the Field metadata in that Table
+  metadata.
+
+  Returns `nil` if no matching metadata could be found."
+  ([metadata :- [:map
+                 [:lib/type [:keyword]]]
+    field    :- [:or ::lib.schema.common/non-blank-string ::lib.schema.id/field]]
```

Comment:
Yes I think so, because if you have a native source query then the correct way to refer to the columns it returns would be something like `[:field "count" {:base-type :type/Integer}]` or whatever -- in that case it's a field without a table

Reply:
OK, so we won't "force" that result to have an alias or sg. similar.